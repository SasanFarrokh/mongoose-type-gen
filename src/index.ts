export * from './interface-generator'
export * from './schema-parser'
