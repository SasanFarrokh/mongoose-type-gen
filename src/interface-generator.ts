import { EOL } from 'os'

const INDENT_CHAR = '    '

export type InterfaceGeneratorTypeDef = { name: string, optional?: boolean, arrayDepth?: number };

export class InterfaceGenerator {
    private deps: InterfaceGenerator[] = []
    private fields = new Map<string, InterfaceGeneratorTypeDef>()

    constructor(
        public readonly name: string,
        public exports: boolean,
        public readonly extend?: string,
    ) {}

    addField(name: string, type: InterfaceGeneratorTypeDef | InterfaceGenerator) {
        if (type instanceof InterfaceGenerator) {
            if (type.name !== this.name) {
                this.deps.push(type)
            }
            this.fields.set(name, { name: type.name })
        } else {
            this.fields.set(name, type)
        }
    }

    addDep(ig: InterfaceGenerator) {
        this.deps.push(ig)
    }

    toString() {
        let result = ''
        for (const dep of this.deps) {
            result += dep.toString()
        }
        result += `${this.exports ? 'export ' : ''}interface ${this.name}${this.extend ? ' extends ' + this.extend : ''} {` + EOL

        result += InterfaceGenerator.fieldsToString(this.fields, INDENT_CHAR, 1)

        result += '}'
        return result.trim() + EOL + EOL
    }

    static fieldsToString(fields: Map<string, InterfaceGeneratorTypeDef>, indent = '', newLine = 1) {
        let result = ''
        const repeatString = (str: string, repeat: number): string => new Array(repeat).fill(str).join('');

        fields.forEach((type, name) => {
            result += `${indent}${name}${type.optional ? '?' : ''}: ${type.name}${repeatString('[]', type.arrayDepth || 0)};` + repeatString(EOL, newLine)
        })
        return result
    }
}
