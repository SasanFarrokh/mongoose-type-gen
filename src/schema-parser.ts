import {Schema, VirtualType} from "mongoose";
import {InterfaceGenerator, InterfaceGeneratorTypeDef} from "./interface-generator";
import {EOL} from "os";

type FieldMap = Map<string, FieldMap | string>

export class SchemaParser {
    private imports: string[];

    constructor(private readonly schema: any, readonly name: string) {
        if (!(schema instanceof Schema)) {
            console.warn('Schema is not instanceof mongoose Schema')
        }
        this.imports = [
            `import { Schema, Model, Document } from 'mongoose' ${EOL + EOL}`
        ]
    }

    fields(tree: Record<string, any>) {
        return Object.entries(tree).filter(([field, type]) => isNaN(+field) && field !== '_id')
    }

    resolveGenerator(tree: Record<string, any>, deps: string[]): InterfaceGenerator {
        const ig = new InterfaceGenerator([this.name].concat(deps).map(el => el[0].toUpperCase() + el.slice(1)).join(''), false)

        this.fields(tree).forEach(([field, type]) => {
            ig.addField(field, this.resolveType(field, type, deps));
        })
        return ig;

    }

    private resolveType(field: string, type: any, deps: string[] = []): InterfaceGeneratorTypeDef | InterfaceGenerator {
        if (this.isVirtualType(type)) {
            // Any for virtual types
            let typeDef: InterfaceGeneratorTypeDef;
            try {
                typeDef = { name: this.resolveNativeType(type.options.type) }
            } catch (err) {
                console.warn(`Virtual field '${field}' is not typed`)
                typeDef = { name: 'any' }
            }
            typeDef.optional = true;
            return typeDef
        }
        else if (this.isNestedObjectType(field, deps, type)) {
            return this.resolveGenerator(type, deps.concat(field))
        }
        else if (this.isNestedSchemaType(type)) {
            return this.resolveGenerator(type.tree, deps.concat(field))
        }
        else if (this.isArrayType(type)) {
            type = this.resolveNativeType(type[0])
            return {
                arrayDepth: 1,
                name: type,
                optional: !type[0].required
            }
        } else {
            return {
                name: this.resolveNativeType(type),
                optional: !type.required
            }
        }
    }
    private resolveNativeType(typeDef: any): string {
        const type = typeDef.type && (typeof typeDef.type === 'function') ? typeDef.type : typeDef;
        switch (true) {
            case type instanceof Schema.Types.ObjectId || type.schemaName === 'ObjectId' || type === 'ObjectId':
                return 'Schema.Types.ObjectId'
            case type === String || type === Schema.Types.String:
                return 'string'
            case type === Number || type === Schema.Types.Number:
                return 'number'
            case type === Boolean || type === Schema.Types.Boolean:
                return 'boolean'
            case type === Date:
                return 'Date'
            case type === Schema.Types.Mixed:
                return 'any'
            case Array.isArray(type):
                return this.resolveNativeType(type[0]) + '[]'
        }
        console.info(typeDef)
        throw new Error('Type is not supported: ' + JSON.stringify(typeDef))
    }

    isVirtualType(fieldConfig: any): boolean {
        return fieldConfig instanceof VirtualType;
    }

    private isNestedObjectType(field: string, deps: string[], fieldConfig: any): boolean {
        return this.schema.nested[deps.concat(field).join('.')] === true
    }

    private isNestedSchemaType(fieldConfig: any): boolean {
        return fieldConfig instanceof Schema;
    }

    private isArrayType(field: any): boolean {
        return Array.isArray(field);
    }

    toString() {
        const rootInterfaceGenerator = this.resolveGenerator(this.schema.tree, []);
        for (const method in this.schema.methods) {
            rootInterfaceGenerator.addField(method, { name: 'Function' })
        }
        rootInterfaceGenerator.exports = true;

        const modelInterface = new InterfaceGenerator(this.name + 'Model', true,`Model<${this.name}Document>`)
        for (const method in this.schema.statics) {
            modelInterface.addField(method, { name: 'Function' })
        }

        const exports = [
            `export type ${this.name}Document = ${this.name} & Document`,
            modelInterface.toString(),
        ]
        return this.imports.join(EOL) + rootInterfaceGenerator.toString() + exports.join(EOL);
    }
}
