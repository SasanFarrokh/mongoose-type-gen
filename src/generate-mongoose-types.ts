import {program} from 'commander';
import * as fs from 'fs';
import * as path from 'path';
import {Schema} from "mongoose";
import {SchemaParser} from "./schema-parser";

program.on('--help', () => {
});

program
    .command('generate [schemaPaths...]')
    .action(async (schemaPaths: string[], cmd: any) => {
        const {outputDir} = cmd;

        // if (!outputDir) {
        //     console.error('An output directory is required. Use -o or --output-dir');
        //     return process.exit(1);
        // }

        if (!schemaPaths || !schemaPaths.length) {
            console.error(
                'No schema files could be found. Please check the required usage by running `ms2tsi -h`',
            );
            return process.exit(1);
        }

        const currentDir = process.env.PWD as string;

        const promises = schemaPaths.map((schemaPath: string) => {
            const schemaFile = require(path.resolve(currentDir, schemaPath));
            for (const key in schemaFile) {
                console.log(`  Generating: ` + schemaPath)
                if (schemaFile[key] instanceof Schema) {
                    const parser = new SchemaParser(schemaFile[key], key)
                    return fs.promises.writeFile(schemaPath.replace(/\.ts/i, '.d.ts'), parser.toString())
                }
            }
        })
        await Promise.all(promises)
    });

program.parse(process.argv);
