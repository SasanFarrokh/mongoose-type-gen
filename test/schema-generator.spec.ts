import {Schema} from 'mongoose'
import * as fs from 'fs'
import {SchemaParser} from "../src/schema-parser";

describe('snapshot test', function () {
    it('should match generated type', function () {

        const family = new Schema({
            familyName: String
        })

        const user = new Schema({
            name: String,
            lastname: {
                type: String,
            },
            rank: Number,
            books: [
                {
                    type: String,
                    default: ['My Book']
                }
            ],
            authId: Schema.Types.ObjectId,
            info: {
                email: String,
                github: {
                    type: String,
                    required: true
                },
                npm: {
                    url: String,
                }
            },
            matrix3D: [[[String]]],
            family: family
        }, {
            timestamps: true
        })

        user.virtual('fullName').get(function (this: any) {
            return this.name + this.lastname
        })

        user.virtual('fullNameTyped', { type: [String] }).get(function (this: any) {
            return this.name + this.lastname
        })

        ;(user.method as any)('writeBook', () => {console.log('writing book')}, {s: true})

        user.static('findUserByBook', function (this: any, book: any) {
            return this.find({ book })
        })

        console.info(user)

        const output = (new SchemaParser(user, 'User')).toString()

        fs.promises.writeFile('./example.ts', output.toString())
        expect(output).toMatchSnapshot();
    });
});

