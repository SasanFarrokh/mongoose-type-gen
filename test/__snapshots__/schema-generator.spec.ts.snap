// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test should match generated type 1`] = `
"import { Schema, Model, Document } from 'mongoose' 

interface UserInfoNpm {
    url?: string;
}

interface UserInfo {
    email?: string;
    github: string;
    npm: UserInfoNpm;
}

interface UserFamily {
    familyName?: string;
    id?: any;
}

export interface User {
    name?: string;
    lastname?: string;
    rank?: number;
    books?: string[];
    authId?: Schema.Types.ObjectId;
    info: UserInfo;
    matrix3D?: string[][][];
    family: UserFamily;
    updatedAt?: Date;
    createdAt?: Date;
    fullName?: any;
    fullNameTyped?: string[];
    initializeTimestamps: Function;
    writeBook: Function;
}

export type UserDocument = User & Document
export interface UserModel extends Model<UserDocument> {
    findUserByBook: Function;
}

"
`;
